### PR TITLE
Fix: Rows with no decisions should be emitted

### PR DESCRIPTION
**Pull request by hudson-newey**
# Fix: Rows with no decisions should be emitted

A short description of the pull request

### Features

None

### Bug Fixes

- Rows with no decisions are now emitted in downloaded results
- Fixes a bug where the first page of a verification grid would really be the second
- Fixes a bug where if there was a decision button without a shortcut, the help dialog would fail to open
- Fixes a bug where the download results button would not should if the `local` attribute was not present

### Code Quality

None

## Related Issues

Fixes: #95 

## Final Checklist

- [x] All commits messages are semver compliant
- [x] Added relevant unit tests for new functionality
- [x] Updated existing unit tests to reflect changes
- [x] Code style is consistent with the project guidelines
- [ ] Documentation is updated to reflect the changes (if applicable)
- [x] Link issues related to the PR
- [x] Assign labels if you have permission
- [ ] Assign reviewers if you have permission
- [ ] Ensure that CI is passing
- [x] Ensure that `pnpm lint` runs without any errors
- [ ] Ensure that `pnpm test` runs without any errors

### CODE REVIEW

**Review by atruskie — APPROVED** [review]
Looks good, minimal changes required

File: src/components/data-source/data-source.ts [hunk 1]
```diff
@@ -69,70 +72,91 @@ export class DataSourceComponent extends AbstractComponent(LitElement) {
...
-
-    // since we do not know the input format of the provided csv or json files
-    // it is possible for users to input a csv file that already has a column
-    // header of "confirmed" or "additional-tags"
-    // to prevent column name collision, we prepend all the fields that we add
-    // to the original data input with "oe"
-    const columnNamespace = "oe-";
+    const results = await this.resultRows();
+    const fileFormat = this.dataFetcher.mediaType ?? "";
+    const fileName = `verified-${this.dataFetcher.file.name}`;
```
[the remark this answers is not in the document]
Comment (hudson-newey) [reply src/components/data-source/data-source.ts:87]: This makes a lot of logical sense, I like it!

File: src/components/data-source/data-source.ts [hunk 2]
```diff
@@ -179,9 +205,14 @@ export class DataSourceComponent extends AbstractComponent(LitElement) {
       throw new Error("for attribute must be set on a data source");
     }
 
-    const dataFetcher = new DataSourceFetcher(this.src);
-    const data = await dataFetcher.json();
-    this.contentType = dataFetcher.contentType;
+    this.dataFetcher = new DataSourceFetcher(this.src);
+    await this.dataFetcher.updateSrc(this.src);
```
[the remark this answers is not in the document]
Comment (hudson-newey) [reply src/components/data-source/data-source.ts]: Fixed in commit [f7c7203](https://github.com/ecoacoustics/web-components/pull/101/commits/f7c7203a792355dd1ac2f56747b31e6697fca108)

File: src/services/dataSourceFetcher.ts [hunk 3]
```diff
@@ -12,58 +11,119 @@ export type DataSourceFetcherContentTypes = "json" | "csv" | "tsv";
 export class DataSourceFetcher extends AbstractDataFetcher {
     public constructor(src: string) {
         super(src);
-        this.src = src;
+        this._src = src;
```
[the remark this answers is not in the document]
Comment (hudson-newey) [reply src/services/dataSourceFetcher.ts]: This was a hack to do with my `AbstractDataFetcher`
I had some grand plans with data fetchers so that pagination, data source fetching, etc... would all use the same base class then implement their own caching strategy

But I didn't end up implementing this functionality

Since `AbstractDataFetcher` is only used by this one class, and its implementation results in me defining two "sources of truth" (one on the DataSourceFetcher, and the AbstractDataFetcher), I've decided to remove the AbstractDataFetcher to make the code easier to read, and less prone to bugs.

Tl;DR - I've removed `_src`, because `AbstractDataFetcher` is no longer needed

File: src/services/dataSourceFetcher.ts [hunk 4]
```diff
@@ -12,58 +11,119 @@ export type DataSourceFetcherContentTypes = "json" | "csv" | "tsv";
...
+     * because json can have an array at the root or an object
+     * we support returning an array or a record
+     * 
+     * @returns the parsed data source as a JSON object or undefined
+     * if the data source could not be fetched
+     */
+    public async json(): Promise<ReadonlyArray<VerificationSubject> | undefined> {
+        return this.jsonModels ??= await this.generateJsonModels();
+    }
+
```
[the remark this answers is not in the document]
Comment (hudson-newey) [reply src/services/dataSourceFetcher.ts]: New name `subjects` and `generateSubjects`

I can use a more specific name now that I don't have to conform to the `AbstractDataFetcher`

File: src/components/verification-grid/help-dialog.ts [hunk 5]
```diff
@@ -74,13 +74,14 @@ export class VerificationHelpDialogComponent extends AbstractComponent(LitElemen
...
-            ${shortcut.key
-          .split("+")
-          .map((key, i, { length }) => html`<kbd class="key">${key}</kbd> ${i !== length - 1 ? "+" : nothing}`)}
-            <span class="description">${shortcut.description}</span>
-          </div>`,
+        ${shortcuts.map((shortcut) => html`
+        <div>
+          ${shortcut.key.split("+")
+        .map((key, i, { length }) => html`
+          <kbd class="key">${key}</kbd> ${i !== length - 1 ? "+" : nothing}
```
[the remark this answers is not in the document]
Comment (hudson-newey) [reply src/components/verification-grid/help-dialog.ts:81]: This is super cool, this is great

image

File: src/models/verification.ts [hunk 6]
```diff
@@ -1,4 +1,14 @@
-export type VerificationSubject = Record<string, unknown>;
+export type VerificationSubjectData = Record<string, unknown>;
+
+export class VerificationSubject {
+  public constructor(data: Record<string, unknown>) {
+    this.data = data;
+    this.identifier = Symbol("verification-subject");
+  }
```
[the remark this answers is not in the document]
Comment (hudson-newey) [reply src/models/verification.ts:7]: When we download results, we read the input file (sequentially), then we can check if there is a decision associated to the row by using the symbol.

Because we read the file sequentially, we can ensure order, even when decisions are made out of order

This might become a problem with random ordering once we implement https://github.com/ecoacoustics/web-components/issues/75, but I'd say that the scope of order tracking is limited to that issue

_Side note_: I found the bug "Fixes a bug where the first page of a verification grid would really be the second" because the order was maintained in the results (I saw that the first 8 rows did not have a decision applied)

[the remark this answers is not in the document]
Comment (atruskie) [reply src/models/verification.ts:7]: OK, we probably need an index tracking variable for order - that will be more reliable. But you're right, out of scope for now
